File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int mx=0;
    void ziggy(TreeNode*t,bool isLeft,int currStep){
        if(!t)return ;
        mx=max(mx,currStep);
        if(isLeft){
            ziggy(t->left,false,currStep+1);
            ziggy(t->right,true,1);
        }
        else{
            ziggy(t->right,true,currStep+1);
            ziggy(t->left,false,1);
        }
    }
    int longestZigZag(TreeNode* root) {
        ziggy(root,true,0);
        ziggy(root,false,0);
        return mx;
    }
};

File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int mx=0;
    public void ziggy(TreeNode t,boolean isLeft,int currStep){
        if(t==null)return ;
        mx=Math.max(mx,currStep);
        if(isLeft==true){
            ziggy(t.left,false,currStep+1);
            ziggy(t.right,true,1);
        }
        else{
            ziggy(t.right,true,currStep+1);
            ziggy(t.left,false,1);
        }
    }
    public int longestZigZag(TreeNode root) {
        ziggy(root,true,0);
        ziggy(root,false,0);
        return mx;
        
    }
}